src/lib.rs

```diff
@@ -39,13 +40,30 @@ pub struct Scope {
     /// Imports
     imports: OrderMap<String, OrderMap<String, Import>>,
 
+    /// The modules defined in this scope.
+    /// 
+    /// Modules are stored in a map rather than in this scope's
+    /// `items` vector so that they may be accessed by name directly
+    /// after being created.
+    modules: OrderMap<String, Module>,
+
     /// Contents of the documentation,
     items: Vec<Item>,
 }
 
 #[derive(Debug, Clone)]
 enum Item {
-    Module(Module),
+    /// A module.
+    /// 
+    /// Unlike all the other variants of `Item`, this does not hold a `Module` 
+    /// directly, but instead holds an `Rc<String>` (the name of the module). 
+    /// This is because the `Module`s defined in a `Scope` do not live in the 
+    /// `Scope`'s `items` vector, but in its `modules` map. When we want to 
+    /// access the `Module` corresponding to this `Item`, we can use the string
+    /// to index the scope's map of modules. This way, modules can be looked up
+    /// by name, but also are defined at ordered locations in the `items` 
+    /// vector.
+    Module(String),
```

Comment:
I'd rather avoid the additional indirection for now.

If doing a sequential scan in the lookup fns is problematic, we can explore replacing `items: Vec<Item>` with an `OrderMap` as order is preserved with that structure as well. This probably should happen in a separate PR.

Reply:
Replacing `items` with an `OrderMap` was the second approach I tried (50b7156f009118e73215d676e45eb47407bf1165), it turned out to be problematic as not all items had obvious keys into that map, and it was difficult to determine keys that wouldn't result in some items being clobbered when they shouldn't be. I'll take another crack at the sequential scan instead.

Reply:
Okay, done in 71348a8737743bd237bcf040461fa742990a6f46.